### PR TITLE
gatk4 markduplicates

### PR DESCRIPTION
**Pull request by ajodeh-juma**
## PR checklist

Closes #XXX 

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [x] If necessary, include test data in your PR.
- [x] Remove all TODO statements.
- [x] Emit the `<SOFTWARE>.version.txt` file.
- [x] Follow the naming conventions.
- [x] Follow the parameters requirements.
- [x] Follow the input/output options guidelines.
- [x] Add a resource `label`
- [x] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [x] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Comment by grst** [comment]
Hi @ajodeh-juma, looks good now! 
You are just missing an entry in the `pytest_software.yml` file to make the pytest-workflow tests run.

**Comment by ajodeh-juma** [comment]
I've just looked into the file `pytest_software.yml` but there seems to be no `gatk4_markduplicates` section but when I generated the module, I got the output as `INFO     Created / edited following files:                                                                                                                           create.py:218
           ./software/gatk4/markduplicates/functions.nf                                                                                                                           
           ./software/gatk4/markduplicates/main.nf                                                                                                                                
           ./software/gatk4/markduplicates/meta.yml                                                                                                                               
           ./tests/software/gatk4/markduplicates/main.nf                                                                                                                          
           ./tests/software/gatk4/markduplicates/test.yml                                                                                                                         
           ./tests/config/pytest_software.yml `

> Hi @ajodeh-juma, looks good now!
> You are just missing an entry in the `pytest_software.yml` file to make the pytest-workflow tests run.

**Comment by grst** [comment]
No idea why it didnt' get updated... please just add the entry manually in that case!

File: software/gatk4/markduplicates/main.nf [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
+    """
+    gatk MarkDuplicates \\
+        --INPUT $bam \\
+        --METRICS_FILE ${prefix}.bam.metrics \\
+        --TMP_DIR . \\
+        --ASSUME_SORT_ORDER coordinate \\
+        --CREATE_INDEX true \\
+        --OUTPUT ${prefix}.md.bam \\
```
Comment (drpatelh) [inline software/gatk4/markduplicates/main.nf:39]:
```suggestion
        --OUTPUT ${prefix}.bam \\
```

File: software/gatk4/markduplicates/main.nf [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    path "*.version.txt"          , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
+    """
+    gatk MarkDuplicates \\
+        --INPUT $bam \\
+        --METRICS_FILE ${prefix}.bam.metrics \\
```
Comment (drpatelh) [inline software/gatk4/markduplicates/main.nf:35]:
```suggestion
        --METRICS_FILE ${prefix}.metrics \\
```

File: software/gatk4/markduplicates/main.nf [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    } else {
+        container "quay.io/biocontainers/gatk4:4.2.0.0--0"
+    }
+
+    input:
+    tuple val(meta), path(bam)
+
+    output:
+    tuple val(meta), path("*.bam"), emit: bam
+    path "*.version.txt"          , emit: version
```
Comment (drpatelh) [inline software/gatk4/markduplicates/main.nf:26]:
```suggestion
    tuple val(meta), path("*.bam")    , emit: bam
    tuple val(meta), path("*.metrics"), emit: metrics
    path "*.version.txt"              , emit: version
```

File: software/gatk4/markduplicates/main.nf [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+
```
Comment (drpatelh) [inline software/gatk4/markduplicates/main.nf:31]:
```suggestion
```

File: software/gatk4/markduplicates/meta.yml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        e.g. [ id:'test', single_end:false ]
+  - version:
+      type: file
+      description: File containing software version
+      pattern: "*.{version.txt}"
+  - bam:
+      type: file
+      description: Marked duplicates BAM file
+      pattern: "*.{bam}"
+
```
Comment (drpatelh) [inline software/gatk4/markduplicates/meta.yml:43]: Add entry for `metrics` output here too

File: tests/software/gatk4/markduplicates/main.nf [hunk 6]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { GATK4_MARKDUPLICATES } from '../../../../software/gatk4/markduplicates/main.nf' addParams( options: [:] )
```
Comment (drpatelh) [inline tests/software/gatk4/markduplicates/main.nf:5]:
```suggestion
include { GATK4_MARKDUPLICATES } from '../../../../software/gatk4/markduplicates/main.nf' addParams( options: [ 'suffix': '.md' ] )
```

File: tests/software/gatk4/markduplicates/test.yml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+- name: gatk4 markduplicates test_gatk4_markduplicates
+  command: nextflow run tests/software/gatk4/markduplicates -entry test_gatk4_markduplicates -c tests/config/nextflow.config
+  tags:
+    - gatk4
+    - gatk4_markduplicates
+  files:
+    - path: output/gatk4/test.md.bam
+      md5sum: b8246f036b9d8a1d0ca7fe34a14ae432
```
Comment (drpatelh) [inline tests/software/gatk4/markduplicates/test.yml:8]: If possible add entry for `metrics` file here too.

File: tests/software/gatk4/markduplicates/main.nf [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+workflow test_gatk4_markduplicates {
+
+    def input = []
+    input = [ [ id:'test', single_end:false ], // meta map
+              file("${launchDir}/tests/data/genomics/sarscov2/bam/test_paired_end.sorted.bam", checkIfExists: true) ]
```
Comment (drpatelh) [inline tests/software/gatk4/markduplicates/main.nf:11]:
```suggestion
    input = [ [ id:'test', single_end:false ], // meta map
              file(params.test_data['sarscov2']['illumina']['test_paired_end_sorted_bam'], checkIfExists: true)
            ]
```

**Review by drpatelh — COMMENTED** [review]
Great work @ajodeh-juma ! Some minor changes. I think you may need to sync your branch with all of the recent commits on `master` to get things going again with the tests.

**Comment by ajodeh-juma** [comment]
> sync your branch with all of the recent commits on `master`

Thanks @drpatelh , I am getting this error: 
```
Removing tests/software/bismark/methylation_extractor/main.nf
Removing tests/software/bismark/genome_preparation/main.nf
Removing tests/data/genomics/sarscov2/gtf/test_genome.gtf.gz
Removing tests/data/genomics/sarscov2/gtf/test_genome.gff3.gz
Removing tests/data/genomics/sarscov2/bed/test2.bed
Removing tests/data/genomics/sarscov2/bam/test_unaligned.bam
Removing tests/data/genomics/sarscov2/bam/test_single_end.sorted.bam
Removing tests/data/genomics/sarscov2/bam/test_single_end.bam
Removing tests/data/genomics/sarscov2/bam/test_paired_end.sorted.bam.bai
Removing tests/data/genomics/sarscov2/bam/test_paired_end.sorted.bam
Removing tests/data/genomics/sarscov2/bam/test_paired_end.bam
Removing tests/data/genomics/sarscov2/bam/test_methylated_paired_end.sorted.bam.bai
Removing tests/data/genomics/sarscov2/bam/test_methylated_paired_end.sorted.bam
Removing tests/data/genomics/sarscov2/bam/test_methylated_paired_end.bam
Removing tests/data/generic/gtf/GCF_000019425.1_ASM1942v1_genomic.gtf
Removing tests/data/generic/fasta/NC_010473.fa.fai
Removing tests/data/generic/fasta/NC_010473.fa
Removing tests/data/generic/fasta/GCF_000019425.1_ASM1942v1_genomic.fna
Auto-merging tests/config/pytest_software.yml
CONFLICT (content): Merge conflict in tests/config/pytest_software.yml
Removing deprecated/shovill/test/nextflow.config
Removing deprecated/shovill/test/main.nf
Removing deprecated/shovill/meta.yml
Removing deprecated/shovill/main.nf
Automatic merge failed; fix conflicts and then commit the result.
```

Yesterday, @grst suggested I add an entry for the tool to the pytest manually and this has been difficult to resolve.

**Comment by ajodeh-juma** [comment]
> Great work @ajodeh-juma ! Some minor changes. I think you may need to sync your branch with all of the recent commits on `master` to get things going again with the tests.

Here are the set of commands I am using

```
git checkout master
git pull
git checkout gatk4-markduplicates
git merge master
```

**Comment by grst** [comment]
To fix the merge conflict, you need to open the conflicting file 
```
CONFLICT (content): Merge conflict in tests/config/pytest_software.yml
```
in an editor (ideally vscode or an other IDE to get some graphical help) and fix the conflicts. 
Then you can `git add` the file and run `git commit` to finish the merge. 

Alternatively, you can discard your local changes by using
```
git checkout master tests/config/pytest_software.yml
```
and then add the entry manually again.

**Comment by drpatelh** [comment]
Yup. We could have fixed the merge conflict via the Github web interface for you but there is no option available above to do this unfortunately. If you are still struggling you could start fresh on another PR? Would mean copying across the stuff you have done here which is pretty much there now. Sorry, alot of things were changing on the modules repo when you created this PR 😏

**Comment by ajodeh-juma** [comment]
> Yup. We could have fixed the merge conflict via the Github web interface for you but there is no option available above to do this unfortunately. If you are still struggling you could start fresh on another PR? Would mean copying across the stuff you have done here which is pretty much there now. Sorry, alot of things were changing on the modules repo when you created this PR 😏

Could you please check whether everything is okay now. The pytest runs although it is running other bunch of tests as well.

File: software/gatk4/markduplicates/meta.yml [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - version:
+      type: file
+      description: File containing software version
+      pattern: "*.{version.txt}"
+  - bam:
+      type: file
+      description: Marked duplicates BAM file
+      pattern: "*.{bam}"
```
Comment (drpatelh) [inline software/gatk4/markduplicates/meta.yml:42]:
```suggestion
  - bam:
      type: file
      description: Marked duplicates BAM file
      pattern: "*.{bam}"
  - metrics:
      type: file
      description: Duplicate metrics file generated by GATK
      pattern: "*.{metrics.txt}"      
```

File: tests/software/gatk4/markduplicates/main.nf [hunk 10]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { GATK4_MARKDUPLICATES } from '../../../../software/gatk4/markduplicates/main.nf' addParams( options: [ 'suffix': '.md' ] )
```
Comment (drpatelh) [inline tests/software/gatk4/markduplicates/main.nf:5]:
```suggestion
include { GATK4_MARKDUPLICATES } from '../../../../software/gatk4/markduplicates/main.nf' addParams( options: [:] )
```

File: tests/software/gatk4/markduplicates/test.yml [hunk 11]
```diff
@@ -0,0 +1,8 @@
+- name: gatk4 markduplicates test_gatk4_markduplicates
+  command: nextflow run tests/software/gatk4/markduplicates -entry test_gatk4_markduplicates -c tests/config/nextflow.config
+  tags:
+    - gatk4
+    - gatk4_markduplicates
+  files:
+    - path: output/gatk4/test.md.bam
```
Comment (drpatelh) [inline tests/software/gatk4/markduplicates/test.yml:7]:
```suggestion
    - path: output/gatk4/test.bam
```

File: software/gatk4/markduplicates/functions.nf [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+                    def ext_list = path_list.collect()
+                    ext_list.add(ext.value)
+                    return "${getPathFromList(ext_list)}/$args.filename"
+                }
+            }
+        } else if (ioptions.publish_files == null) {
+            return "${getPathFromList(path_list)}/$args.filename"
+        }
+    }
+}
```
Comment (drpatelh) [inline software/gatk4/markduplicates/functions.nf:60]:
```suggestion
}

```

File: software/gatk4/markduplicates/meta.yml [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  - metrics:
+      type: file
+      description: Duplicate metrics file generated by GATK
+      pattern: "*.{metrics.txt}"
```
Comment (drpatelh) [inline software/gatk4/markduplicates/meta.yml:46]:
```suggestion
      pattern: "*.{metrics.txt}"
```

**Comment by drpatelh** [comment]
Ok. Sorted out the merge conflicts and tried to get the tests passing. Let's see what happens...

**Review by drpatelh — APPROVED** [review]

**Pull request merged**